Cloudwatch: Fix Unexpected error

sarahzinger opened the pull request:
**What is this feature?**

Fixes issue where an unparsed error is thrown in Cloudwatch when a datasource is created:
<img width="523" alt="Screenshot 2023-09-05 at 4 15 36 PM" src="https://github.com/grafana/grafana/assets/6620164/375c3f20-ea9f-4506-8ed7-97ba3f1ee8e5">

In order to address fixing this, I ended up refactoring the /regions endpoint to follow the same architecture we now have in most of cloudwatch resource routes. This helps to ensure we return a well formatted error response like so:
<img width="768" alt="Screenshot 2023-09-05 at 4 17 43 PM" src="https://github.com/grafana/grafana/assets/6620164/86f049f7-569b-4969-b03b-ebab587e21b6">

rather than this string which our global error handling does not parse well:
<img width="511" alt="Screenshot 2023-09-05 at 4 18 01 PM" src="https://github.com/grafana/grafana/assets/6620164/2890f764-2b5c-499d-82d9-db3faf92468b">

Arguably there are a few reasons this error is happening, and many of them are worth tackling. Some of them I've covered in this pr and some of which I did not, figuring we could address it later:
1) (not addressed) when regions doesn't have auth session yet it returns an error rather than our fallback list of regions. It is debatable to me what the expected behavior should be in this instance. Perhaps this is as desired...but then what's the point of fallback regions if we're not going to use them? We end up having fallback regions in our frontend as well as in our backend which seems wrong. I think probably we should someday make this return the fallback rather than an error, but that felt like something we could work on in the future.
2) (not addressed) our UI is making a request for /regions before we have an auth saved, even though the frontend already has a default list of regions it can use (located in the grafana-aws-sdk). Again debatable to me what the behavior here should be. If we remove the error entirely and always use the fallback list, then it seems reasonable to me that we make this request right away. But if we're going to throw an error, then it seems strange, like we should wait until there is a saved configuration. But again I think probably it's fine as is once we stop returning errors. 
3) (addressed) we were trying to create a session with a default region of "", rather than throwing an error. This led to a confusing error message, now we return an error that clearly states we're trying to create an auth session with a default region. 
4) (addressed) we were return a string rather than an error object, which our frontend could not parse. We still throw the error but it is no longer noticable unless you have the network tab open. In my opinion this is a good reminder that we need to revisit error handling in Cloudwatch in general. 

For such a tiny bug, I ended up re-writing a lot of code, because it felt like a good opportunity to clean up some tech debt and get better familiar with our backend architecture. But it also made me a bit concerned that I might have unexpected changes here, so I went ahead and put the new stuff behind a feature flag. I feel reasonably confident with it so I was going to have it enabled by default, but if it seems to be causing issues we'll have a quick way to turn it off without reverting the commit. 

An interesting note here is that I saw we have a `regionsCache` which I've removed. I realized while trying to add it back, that we're caching regions under a key of `Profile` which is often undefined, which seems wrong. I think it's not unreasonable for us to add in a cache and create a proper cache key (maybe out of a hash of the auth settings the way we do with settings?) but for now since it seems it was often unused, I've just removed it, but let me know if I'm missing anything!

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/73294

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

sarahzinger committed:
Fix Unexpected error

sarahzinger committed:
Update comment

sarahzinger commented on pkg/tsdb/cloudwatch/client_factory.go:
```diff
@@ -52,8 +52,8 @@ var NewCWLogsClient = func(sess *session.Session) cloudwatchlogsiface.CloudWatch
-	return ec2.New(provider)
+var NewEC2Client = func(sess *session.Session) models.EC2APIProvider {
+	return ec2.New(sess)
 }
```
this change feels the shadiest to me. I don't really understand how this is so easily changed without affecting things? maybe it will become more apparent in tests?

sarahzinger replied:
yeah I think it doesn't need to change, I don't really know why I did this, putting it back.

sarahzinger commented on pkg/tsdb/cloudwatch/models/http_error.go:
```diff
@@ -8,6 +8,7 @@ type HttpError struct {
 	StatusCode int
 }
 
+// Note: an HttpError is a user facing error. It should not contain any sensitive information.
```
this is more aspirational than anything. Maybe it doesn't really belong in this pr

fridgepoet commented on pkg/tsdb/cloudwatch/routes/regions.go:
```diff
@@ -0,0 +1,45 @@
+	service, err := newRegionsService(ctx, pluginCtx, reqCtxFactory, defaultRegion)
+
+	if err != nil {
+		return nil, models.NewHttpError("Unexpected error while fetching regions", http.StatusInternalServerError, err)
+	}
+
+	regions, err := service.GetRegions()
+	if err != nil {
+		return nil, models.NewHttpError("Unexpected error while fetching regions", http.StatusInternalServerError, err)
+	}
+
+	regionsResponse, err := json.Marshal(regions)
+	if err != nil {
+		return nil, models.NewHttpError("Unexpected error while fetching regions", http.StatusInternalServerError, err)
+	}
```
Probably make these error strings different so that log debugging can identify just one path

sarahzinger committed:
Change newec2Client factory back to how it was.

sarahzinger committed:
Lint

sarahzinger committed:
Change up error messages to be easier to differentiate

sarahzinger committed:
Remove comment for now about http error, will revisit error handling later

sarahzinger committed:
Merge main, resolve conflicts

sarahzinger committed:
Fix existing tests

sarahzinger committed:
remove junk log lines and fix integration test setup

sarahzinger committed:
Fix log actions tests

sarahzinger committed:
Add tests for routes and services

sarahzinger committed:
Fix test setup

sarahzinger commented:
Just adding some notes for myself here so I don't forget:
- Right now I'm changing the backend response from SuggestData to a ResourceResponse, need to change the frontend to handle this.
- I removed the regionsCache for simplicity when moving the resource to the right handler, but I think I should add it back, since there's no reason to make users repeatedly ping for regions all the time. 
-  this is feeling pretty large for such a tiny bug. I'm thinking about adding a feature flag
- maybe I can also updated the frontend to not ping at all when not necessary.

sarahzinger committed:
Put new regions behind feature flag

sarahzinger committed:
Merge conflicts

sarahzinger committed:
Add test to backend for feature gating

sarahzinger committed:
Add back old tests for now, can remove once we remove feature flag

sarahzinger committed:
Fix test

sarahzinger committed:
Merge branch 'main' into sarahzinger/fix-unexpected-error

fridgepoet commented on pkg/tsdb/cloudwatch/cloudwatch_integration_test.go:
```diff
@@ -239,4 +250,43 @@ func Test_CloudWatch_CallResource_Integration_Test(t *testing.T) {
 		assert.JSONEq(t, `[{"value":{"name":"field1","percent":50}},{"value":{"name":"field2","percent":50}}]`, string(sent.Body))
 	})
+
+	t.Run("Should handle region requests and return regions from the api", func(t *testing.T) {
```
What are we showing in this test?
(Are we showing that with the new refactored endpoint, now the returned JSON is in a different format?) What's the argument for the new format again?

Why do we only assert one of the regions is present? Why not assert the entire body?

sarahzinger replied:
> What are we showing in this test

To me our integration tests are showing that when you put together a new executor it will handle requests for /regions. We are also specifying that when the feature flag is set to true, it will send back regions in a particular format. I think all of our integration tests here are mostly demonstrating like a happy path of what endpoints our backend handles, to just quickly validate that our models, services, and routes handlers all kind of work together.

> (Are we showing that with the new refactored endpoint, now the returned JSON is in a different format?) What's the argument for the new format again?
Yes.

The new format is the same as the other resource formats. Before it sent things over like this: 
![Screenshot 2023-09-19 at 12 52 24 PM](https://github.com/grafana/grafana/assets/6620164/c0562678-7aa3-42bd-8bd6-c46c8df19b89) 
to something like this:
![Screenshot 2023-09-19 at 12 55 01 PM](https://github.com/grafana/grafana/assets/6620164/27ea87c5-5d44-48d3-99eb-ac4522d5aedc)

This also follows the changes we did with log groups:
![Screenshot 2023-09-19 at 12 57 19 PM](https://github.com/grafana/grafana/assets/6620164/6fe7c15c-13df-4cd7-a38a-742b6fd4c85b)

The reasoning here is that the former is very tied to the UI element and kind of brittle as a result. The latter is hopefully a bit more flexible as ui's change, and more accurately reflects the actual data. 

I think arguably this change does not HAVE to happen now just to fix the unexpected token bug, but I think we might as well kind of change this all at once? I think it was going to be just as much work imo to do it the "right" way as it was to keep the old stuff so I felt like this seemed ok to me. 

> Why do we only assert one of the regions is present? Why not assert the entire body?
It just seemed easier to read and to write. The Regions endpoint will ultimately return different regions depending on your particular aws setup, so it seemed reasonable to me to just validate that it is returning in the format we expect.

fridgepoet commented on pkg/tsdb/cloudwatch/cloudwatch_integration_test.go:
```diff
@@ -47,7 +54,11 @@ func Test_CloudWatch_CallResource_Integration_Test(t *testing.T) {
-		return DataSource{Settings: models.CloudWatchSettings{}}, nil
+		return DataSource{Settings: models.CloudWatchSettings{
+			AWSDatasourceSettings: awsds.AWSDatasourceSettings{
+				Region: "us-east-1",
```
There's something about this "us-east-1" here and the assertions below that may be misleading.
For example you can put any string like "some default region" here and the tests will still pass because the regions are coming from a constant list.
What could be interesting would be to show that when Region is omitted here (len == 0 or Region == ""), then you assert a 400 status, bad request. May be able to assert the error message as well?
I believe this is what your PR adds.

sarahzinger replied:
We have that captured in Regions_test but I can put that here too. Wasn't sure if we only test happy path use cases for our integration tests...but that seems silly, we should test for known errors!

fridgepoet commented on pkg/tsdb/cloudwatch/cloudwatch_test.go:
```diff
@@ -111,7 +113,11 @@ func Test_CheckHealth(t *testing.T) {
 	t.Run("successfully query metrics and logs", func(t *testing.T) {
 		client = fakeCheckHealthClient{}
 		im := datasource.NewInstanceManager(func(s backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
```
Not a big deal, wondering here if we shouldn't just declare `im` once near the top of `Test_CheckHealth` since it has become a bit big. Can improve the focus and readability of the check health tests.

fridgepoet commented on pkg/tsdb/cloudwatch/client_factory.go:
```diff
@@ -49,10 +49,10 @@ var NewCWLogsClient = func(sess *session.Session) cloudwatchlogsiface.CloudWatch
 //
 // Stubbable by tests.
-var newEC2Client = func(provider client.ConfigProvider) models.EC2APIProvider {
+var NewEC2Client = func(provider client.ConfigProvider) models.EC2APIProvider {
```
How come the signature isn't `func(sess *session.Session) models.EC2APIProvider` like the other refactors?

sarahzinger replied:
I realized we didn't need to change this to get it to work so I reverted it. When I look at the documentation they describe the first arg as a clientProvider so I kept it, although in the example they pass a session? https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#New I guess sessions implement clientProviders? Felt like if I didn't have to change this I shouldn't, but open to your thoughts!

fridgepoet replied:
Can't think off the top of my mind why we had it the other way. Now it seems to be a mix. I see below it's this way, too. Doesn't seem like a big deal. Thanks for having a look!

fridgepoet commented on pkg/tsdb/cloudwatch/cloudwatch_test.go:
```diff
@@ -194,12 +212,17 @@ func TestQuery_ResourceRequest_DescribeLogGroups_with_CrossAccountQuerying(t *te
 	NewMetricsAPI = func(sess *session.Session) models.CloudWatchMetricsAPIProvider { return nil }
 	NewOAMAPI = func(sess *session.Session) models.OAMAPIProvider { return nil }
+	NewEC2Client = func(provider awsclient.ConfigProvider) models.EC2APIProvider {
+		return fakeEC2Client{}
```
I think this can also `return nil` which is to say DescribeLogGroups doesn't need EC2 as a dependency

fridgepoet commented:
Just some thoughts and totally defer to your judgement: 
Pretty large number of changes and I feel a bit surprised that we feel a need for a feature flag. I wonder if we would have been able to break this up and felt more comfortable about delivering a smaller set of changes a bit at a time?

fridgepoet commented on pkg/tsdb/cloudwatch/routes/regions_test.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+func TestRegionsRoute(t *testing.T) {
+	reqCtxFunc := func(_ context.Context, pluginCtx backend.PluginContext, region string) (reqCtx models.RequestContext, err error) {
```
This doesn't seem to be needed -- I think we're at a lower level here. You can replace it with `nil` in the lines below.
```suggestion
```

fridgepoet commented on pkg/tsdb/cloudwatch/services/regions_test.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/stretchr/testify/mock"
+)
+
+type EC2Mock struct {
+	mock.Mock
+}
+
+func (e *EC2Mock) DescribeRegions(in *ec2.DescribeRegionsInput) (*ec2.DescribeRegionsOutput, error) {
+	args := e.Called()
+	mockResponse := args.Get(0)
+	if mockResponse == nil {
+		return nil, args.Error(1)
+	}
+	return mockResponse.(*ec2.DescribeRegionsOutput), args.Error(1)
+}
+
```
This type can be in `mocks` if you want to match the refactoring.
Any reason for the logic in the method there?

sarahzinger replied:
cool moving it to mocks, sadly I get the same nil error as above if I remove the conditional :/

sarahzinger committed:
Remove log line

Co-authored-by: Shirley <4163034+fridgepoet@users.noreply.github.com>

fridgepoet commented on pkg/tsdb/cloudwatch/cloudwatch_integration_test.go:
```diff
@@ -28,17 +31,21 @@ func Test_CloudWatch_CallResource_Integration_Test(t *testing.T) {
 	}
-
+	NewEC2Client = func(provider client.ConfigProvider) models.EC2APIProvider {
+		return fakeEC2Client{}
```
Sorry, missed this earlier, but not a big deal if you leave it this way.

This is the "new" mock client which uses testify/mock. Eventually it would be great to replace all of the "old" one and only have this mock left. You'll have to add the `".On("DescribeRegions"....`

```suggestion
		return &mockEC2Client{}
```

In the cases where there is the "old" mock and a "new" one, we could have marked the "old" as deprecated. Not sure why we didn't think of that before...

sarahzinger commented on pkg/tsdb/cloudwatch/models/api.go:
```diff
@@ -21,8 +22,10 @@ type RequestContext struct {
+	EC2APIProvider        EC2APIProvider
 	Settings              CloudWatchSettings
 	Features              featuremgmt.FeatureToggles
+	RegionCache           *sync.Map
```
So it is technically used in the old path here: https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/metric_find_query.go#L45

I'm using a feature flag to preserve the old path just in case I'm misunderstanding something and we want to keep the old regions handler while we revert/fix the change. 

When I look at the code it seems to only cache the region on instance.Settings.Profile as a cachekey but my understanding is that instance.Settings.Profile is often blank, IIRC it's only set when using the credentials file provider? 

It kind of feels to me like if we only sometimes use a regions cache, it's probably not necessary to preserve. If we someday decide we're making too many api calls to fetch region it feels like the kind of thing we could add back a cache for and ensure it's used regardless of authentication method. 

But again it's totally possible I'm misreading this or something so I wanted the option to keep the old behavior if it turns out there is something amiss, so I added back the RegionCache but only use it if the feature flag is set to off.

fridgepoet replied:
[This regionCache](https://github.com/grafana/grafana/blob/dd8f88b194e0aedd98e5b51ced7a088e03681f6f/pkg/tsdb/cloudwatch/cloudwatch.go#L116) is what I'm seeing being used still.

The one "RegionCache" seems to have been introduced here and seems to be unused --maybe an artifact from some refactoring? Let me know if I have missed something!

sarahzinger replied:
yikes my bad thank you for spotting!

sarahzinger commented on pkg/tsdb/cloudwatch/routes/regions_test.go:
```diff
@@ -0,0 +1,99 @@
+
+func (r *RegionsService) GetRegions() ([]resources.ResourceResponse[resources.Region], error) {
+	args := r.Called()
+	mockResponse := args.Get(0)
+	if mockResponse == nil {
+		return nil, args.Error(1)
+	}
```
if I remove it I get 
```panic: interface conversion: interface {} is nil, not []resources.ResourceResponse[github.com/grafana/grafana/pkg/tsdb/cloudwatch/models/resources.Region]
```

is there a better way I'm missing maybe?

fridgepoet replied:
Ok I see. This is just a little type issue with the type of `nil` which is interface{} is the .Return and what the args.Get(0) is typed to return here.
Just need to make the type of `nil` in Return match the type of args.Get(0).
Will put suggestions.

sarahzinger committed:
Fixes after CR

fridgepoet commented on pkg/tsdb/cloudwatch/mocks/regions.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (r *RegionsService) GetRegions() ([]resources.ResourceResponse[resources.Region], error) {
+	args := r.Called()
+	mockResponse := args.Get(0)
+	if mockResponse == nil {
+		return nil, args.Error(1)
+	}
+	return args.Get(0).(([]resources.ResourceResponse[resources.Region])), args.Error(1)
```
```suggestion
	args := r.Called()
	return args.Get(0).([]resources.ResourceResponse[resources.Region]), args.Error(1)
```

fridgepoet commented on pkg/tsdb/cloudwatch/mocks/regions.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+func (e *EC2Mock) DescribeRegions(in *ec2.DescribeRegionsInput) (*ec2.DescribeRegionsOutput, error) {
+	args := e.Called()
+	mockResponse := args.Get(0)
+	if mockResponse == nil {
+		return nil, args.Error(1)
+	}
+	return mockResponse.(*ec2.DescribeRegionsOutput), args.Error(1)
```
```suggestion
	args := e.Called()
	return args.Get(0).(*ec2.DescribeRegionsOutput), args.Error(1)
```

fridgepoet commented on pkg/tsdb/cloudwatch/services/regions_test.go:
```diff
@@ -0,0 +1,44 @@
+
+	t.Run("forwards error if DescribeRegions errors out", func(t *testing.T) {
+		ec2Mock := &mocks.EC2Mock{}
+		ec2Mock.On("DescribeRegions").Return(nil, assert.AnError)
```
```suggestion
		ec2Mock.On("DescribeRegions").Return((*ec2.DescribeRegionsOutput)(nil), assert.AnError)
```

fridgepoet commented on pkg/tsdb/cloudwatch/routes/regions_test.go:
```diff
@@ -0,0 +1,84 @@
+		newRegionsService = func(_ context.Context, pluginCtx backend.PluginContext, reqCtxFactory models.RequestContextFactoryFunc, region string) (models.RegionsAPIProvider, error) {
+			return &mockRegionService, nil
+		}
+		mockRegionService.On("GetRegions", mock.Anything).Return(nil, errors.New("aws is having some kind of outage")).Once()
```
```suggestion
		mockRegionService.On("GetRegions", mock.Anything).Return([]resources.ResourceResponse[resources.Region](nil), errors.New("aws is having some kind of outage")).Once()
```

sarahzinger committed:
Fixes after CR

sarahzinger committed:
Merge branch 'main' into sarahzinger/fix-unexpected-error

sarahzinger committed:
Missed a few merge conflict fixes

sarahzinger committed:
Use new mocks or specify when using old mocks
